plugins/hls-change-type-signature-plugin/src/Ide/Plugin/ChangeTypeSignature.hs

```diff
@@ -0,0 +1,247 @@
+diagnosticToChangeSig decls diagnostic = do
+    -- regex match on the GHC Error Message
+    (expectedType, actualType', declName) <- matchingDiagnostic diagnostic
+    let actualType = tidyTypeSignature actualType'
+    -- Find the definition and it's location
+    (declSrcSpan, ghcSig) <- findSigLocOfStringDecl decls (T.unpack declName)
+    -- Make sure the given "Actual Type" is a full signature
+    isValidMessage expectedType ghcSig
+    pure $ ChangeSignature{..}
+
+-- | Does the GHC Error Message give us a Signature we can use?
+-- We only want to change signatures when the "expected signature" given by GHC
+-- matches the one given by the user
+isValidMessage :: DefinedSig -> ExpectedSig -> Maybe ()
+isValidMessage defSig expSig = if defSig == expSig then Just () else Nothing
+
+-- | If a diagnostic has the proper message create a ChangeSignature from it
+matchingDiagnostic :: Diagnostic -> Maybe (ExpectedSig, ActualSig, DeclName)
+matchingDiagnostic diag@Diagnostic{_message} = case map (unwrapMatch . (=~) _message) errorMessageRegexes of
+                                                       []  -> Nothing
+                                                       css -> join $ find isJust css
+    where
+        unwrapMatch :: (Text, Text, Text, [Text]) -> Maybe (ExpectedSig, ActualSig, DeclName)
+        -- due to using (.|\n) in regex we have to drop the erroneous, but necessary ("." doesn't match newlines), match
+        unwrapMatch (_, _, _, [exp, act, _, name]) = Just (exp, act, name)
+        unwrapMatch _                              = Nothing
+
+-- | List of regexes that match various Error Messages
+errorMessageRegexes :: [Text]
+errorMessageRegexes = [ -- be sure to add new Error Messages Regexes at the bottom to not fail any existing tests
+    "Expected type: (" <> typeSigRegex <> ")\n +Actual type: (" <> typeSigRegex <> ")\n(.|\n)+In an equation for ‘(.+)’"
+    , "Couldn't match expected type ‘(" <> typeSigRegex <> ")’ with actual type ‘(" <> typeSigRegex <> ")’\n(.|\n)+In an equation for ‘(.+)’"
+    ]
+    where
+        -- The brackets in position 1 and 2 MUST BE IN THAT ORDER
+        -- Any other combination of brackets seems to fail
+        typeSigRegex = "[][a-zA-Z0-9 ->]+"
+
+-- | Given a String with the name of a declaration, find that declarations and give back
+-- the type signature location and the full signature
+-- This is a modified version of functions found in Development.IDE.Plugin.CodeAction
+-- This function returns the actual location of the signature rather than the actual signature
+-- We also don't have access to `fun_id` or other actual `id` so we must use string compare instead
+findSigLocOfStringDecl :: SigName => [LHsDecl GhcPs] -> String -> Maybe (RealSrcSpan, DefinedSig)
+findSigLocOfStringDecl decls declName = do
+    -- can we simplify this logic? Just want to make sure ghcSig is a Just value
+    (RealSrcSpan rss _, Just ghcSig) <- listToMaybe [ (locA ss, sigToText ts)
```

Comment:
Ugh, this is a partial pattern match, does this call `fail` or something if it doesn't match? ugly.

Reply:
At least don't do the match here and add a `ghcSig <- mayGhcSig` line below or something. Haven't processed the rest enough to think of a simplification.